app/controllers/stripe-setup/director/post.controller.js

```diff
@@ -0,0 +1,168 @@
+  const currentCredential = getCurrentCredential(req.account)
+
+  const stripeAccountSetup = req.account.connectorGatewayAccountStripeProgress
+
+  if (!stripeAccountSetup) {
+    return next(new Error('Stripe setup progress is not available on request'))
+  }
+  if (stripeAccountSetup.director) {
+    req.flash('genericError', 'You’ve already provided director details. Contact GOV.UK Pay support if you need to change them.')
+    return res.redirect(303, formatAccountPathsFor(paths.account.dashboard.index, req.account.external_id))
+  }
+
+  const formFields = getFormFields(req.body, listOfFields)
+  const errors = validateDirector(req.body, formFields)
+
+  if (!lodash.isEmpty(errors)) {
+    const pageData = {
+      firstName: formFields[FIRST_NAME_FIELD],
+      lastName: formFields[LAST_NAME_FIELD],
+      email: formFields[EMAIL_FIELD],
+      dobDay: formFields[DOB_DAY_FIELD],
+      dobMonth: formFields[DOB_MONTH_FIELD],
+      dobYear: formFields[DOB_YEAR_FIELD]
+    }
+
+    pageData['errors'] = errors
+
+    return response(req, res, 'stripe-setup/director/index', {
+      ...pageData, isSwitchingCredentials, collectingAdditionalKycData, currentCredential
+    })
+  } else {
+    try {
+      const stripeAccountId = await getStripeAccountId(req.account, isSwitchingCredentials, req.correlationId)
+      const personsResponse = await listPersons(stripeAccountId)
+      const data = lodash.get(personsResponse, 'data')
+      const director = data !== undefined ? data.filter(person => person.relationship && person.relationship.director === true).pop() : undefined
+
+      if (director && director.id) {
+        await updateDirector(stripeAccountId, director.id, buildStripePerson(formFields))
+      } else {
+        await createDirector(stripeAccountId, buildStripePerson(formFields))
+      }
+
+      await updateCompany(stripeAccountId, { directors_provided: true })
+      await connector.setStripeAccountSetupFlag(req.account.gateway_account_id, 'director', req.correlationId)
+
+      logger.info('Director details submitted for Stripe account', {
+        stripe_account_id: stripeAccountId,
+        is_switching: isSwitchingCredentials,
+        collectingAdditionalKycData: collectingAdditionalKycData
+      })
+      if (isSwitchingCredentials) {
+        return res.redirect(303, formatAccountPathsFor(paths.account.switchPSP.index, req.account.external_id))
+      } else if (collectingAdditionalKycData) {
+        req.flash('generic', 'Details of director successfully completed')
+        return res.redirect(303, formatAccountPathsFor(paths.account.yourPsp.index, req.account && req.account.external_id, currentCredential.external_id))
+      } else {
+        return res.redirect(303, formatAccountPathsFor(paths.account.stripe.addPspAccountDetails, req.account && req.account.external_id))
+      }
+    } catch (err) {
+      next(err)
+    }
+  }
+}
+
+async function getStripeAccountId (account, isSwitchingCredentials, correlationId) {
```

Comment:
Could this be shared as a utility functions, as I'm sure we do this in a lot of places?

Reply:
yes, going to introduce utility and refactor separately.